packager/hls/base/media_playlist.cc

```diff
@@ -345,6 +368,8 @@ MediaPlaylist::MediaPlaylist(const HlsParams& hls_params,
       name_(name),
       group_id_(group_id),
       media_sequence_number_(hls_params_.media_sequence_number) {
+        start_timestamp = base::Time::Now().ToDoubleT();
```

Comment:
It should be generated in the master playlist and passed to the media playlist otherwise there will be inconsistencies among the the media playlists.

Reply:
Master list doesn't create any MediaPlaylist-s, only manipulates them in list(s) and, finally, writes all to file.
Wherein MediaPlaylist can be used without MasterPlaylist and, therefore, must generate the timestamp inside itself.
Thus, if we want to generate timestamps in MasterPlaylist, then we must provide some offset for every MediaPlaylist and all it's segments when MasterPlaylist writing to file. It looks like a dirty hack for me.
SimpleHlsNotifier, on the contrary, creates list(s) of MediaPlaylist-s and thus, I think, can be used for the timestamp generation.
is it correct?

Reply:
I don't think that is correct. Let me clarify what I mean.

Per HLS specification:

>    o  If any Media Playlist in a Master Playlist contains an EXT-X-
      PROGRAM-DATE-TIME tag, then all Media Playlists in that Master
      Playlist MUST contain EXT-X-PROGRAM-DATE-TIME tags with consistent
      mappings of date and time to media timestamps.

The date and time must be consistent for all media playlists in the master playlist. 

However, in your implementation, the |start_timestamp| is re-calculated for every media playlist. They will not be consistent, as different media playlists are created at different time. Although the difference may be small in most cases, likely negligible in some cases, but it will be significant sometimes.

To ensure a consistent date time mapping, there should only be one source of date time mapping. The best approach is to have a single mapping of media timestamp to absolute date/time, which is then propagate to all media playlists. One way to achieve that is to use master playlist to coordinate the mapping. Of course, there are other ways to do it.

Reply:
Yes, specification says that "the date and time must be consistent for all media playlists in the master playlist". And the current solution does not contradict this. Specification does not say that "the date and time must be consistent ACROSS all media playlists in the master playlist". But even so: for example, one stream came (started) immediately, another after 678 ms, the third after 5 s, the fourth after 1.5 s, etc; how to make their timestamps consistent with each other if the streams doesn't reference each other in any way? Do you want to make their start timestamps are the same? Then content will be corrupted, and that's not correct. But if not (let's leave individual start timestamps) - every media playlist initializes its start timestamps upon creating, this happens immediately when the corresponding stream starts.
On the other hand, we're getting current system date and time to initialize start timestamps. This is a really single source of date and time mapping.
Please explain if I misunderstood.

Reply:
It seems that the main problem is that you disagree that the date/time to media timestamp mapping has to be consistent across ALL media playlists.

> If any Media Playlist in a Master Playlist contains an EXT-X-
PROGRAM-DATE-TIME tag, then all Media Playlists in that Master
Playlist MUST contain EXT-X-PROGRAM-DATE-TIME tags with consistent
mappings of date and time to media timestamps.

It mentions clearly that "all media playlists must ... with consistent mappings of date and time to media timestamps".

Do you still have a different interpretation of the statement?

Reply:
Yes.
You said "across ALL". But specification NOT.
I asked you "please explain if I misunderstood" - I meant "describe how you imagine the result", but not to simply repeat your previous post.
And did you really read my explanations? Where the mistake?
Anyway, how should what you want look like? Should all streams (media playlists) have the same start timestamps regardless of the arrival time? I'm really trying to understand.

Reply:
Ping